LandSandBoat/server
# [LUA] Create job_utils/paladin, migrate job functionality, and update JAs

**_I affirm:_**

## What does this pull request do?



**Job Abilities**
- Create job_utils/paladin.lua and populates necessary ability check and ability use functions.
- Moves functionality from all Paladin globals/abilities files to globals/job_utils/paladin.lua.
- Sets up "Enhances 'Divine Emblem' effect" mod to adjust Divine Emblem enmity bonus.
- Sets up Intervene to check for equipped shield, do damage, and inflict Intervene status effect on target.
- Sets up Majesty to apply Cure Potency II and Cure Recast mods, and make Cure and Protect spells AoE.
- Sets up Sepulcher to be usable against Undead enemies only.
- NOTE: Additional changes will be added to this list as they occur.

**Effects**
- Converts Fealty to check resistance inside the globals/magic.lua applyResistanceEffect function.
- Updates Fealty status effect to be dispelable.
- Sets up the Intervene status effect and debuffs.
- Sets up the Majesty status effect and cure potency/white magic recast bonuses.
- Sets up the Palisade status effect and shield block bonus.
- Converts Rampart to SDT and adds the Rampart "Enhances 'Iron Will' effect" fast cast bonus.
- Converts Reprisal effect bonuses to the core battleutils.cpp GetBlockRate function.
- Reduces the enmity bonus from Sentinel by 50 if Paladin is not the player's main job.
- Sets up the Sepulcher status effect and debuffs.

**Augments**
- Sets up "Enhances 'Iron Will' effect" mod and gear augment (Valor/Caballarius Coronet).
- Sets up "Enhances 'Fealty' effect" mod and gear augment (Valor/Caballarius Surcoat).
- Sets up "Enhances 'Chivalry' effect" mod and gear augment (Valor/Caballarius Gauntlets).
- Sets up "Enhances 'Invincible' effect" mod and gear augment (Valor/Caballarius Breeches).
- Sets up "Enhances 'Guardian' effect" mod and gear augment (Valor/Caballarius Leggings).

**Core/LUA Functions**
- Adds lua_spell.cpp getSpellFamily function for pulling a spell's SPELLFAMILY.
- Adds battleutils.cpp GetSpellAoEType "Caster in party" check for Majesty and Accession.
- Adds spell.cpp getAoEMessage single target -> AoE "recovers HP" message override.

## References

All job abilities - https://www.bg-wiki.com/ffxi/Paladin
Shield Block Rate - https://www.bg-wiki.com/ffxi/Shield_Skill

## Steps to test these changes

1. Upload changes to your server.
2. Load the server.
3. Use the `!changejob PLD 99` command.
4. Use all JA's to confirm they do not throw errors.
5. Test all JA's and compare their behavior to BGWiki JA descriptions.
